src/uchen/tboard/tboard_file.cc

```diff
@@ -2,24 +2,167 @@
+  if (path.empty()) {
+    return absl::InvalidArgumentError("Log file name must be provided");
+  }
+
+  if (path.find(uchen::tboard::kLogKey) == std::string::npos) {
+    return absl::InvalidArgumentError(absl::StrFormat(
+        "Log file name must contain \"%s\" to be processed by TensorBoard.",
+        uchen::tboard::kLogKey));
+  }
+
+  // If the path is a directory, make the directory if it doesn't exist.
+  if (path.find('/') != std::string::npos) {
+    std::string_view directory = path.substr(0, path.find_last_of('/'));
+
+    if (!std::filesystem::exists(directory)) {
+      if (!std::filesystem::create_directories(directory)) {
+        return absl::InternalError("Failed to create logs directory");
+      }
+    }
+  }
+
   std::ofstream file(std::string(path), std::ios::binary);
   if (!file.is_open()) {
     return absl::InternalError("Failed to open file");
   }
   return TBoardFile(std::move(file));
 }
 
-void TBoardFile::RecordLoss(float loss) {
+void TBoardFile::RecordLoss(double loss, uint32_t step) {
   LOG(INFO) << "Recording loss: " << loss;
-  file_ << "Loss: " << loss << std::endl;
+  TBoardFile::AddScalar("loss", loss, step);
+}
+
+void TBoardFile::RecordScalar(const std::string& tag, double value,
+                              uint32_t step) {
+  TBoardFile::AddScalar(tag, value, step);
+}
+
+void TBoardFile::RecordImage(const std::string& tag, const std::string& image,
+                             uint32_t step, uint32_t width, uint32_t height,
+                             uint32_t channel, const std::string& image_name,
+                             const std::string& image_description) {
+  std::string encoded_image = EncodeImage(image);
+
+  LOG(INFO) << "Recording image: " << tag;
+
+  tensorflow::SummaryMetadata* metadata = new tensorflow::SummaryMetadata();
+  if (!image_name.empty()) {
+    metadata->set_display_name(image_name);
+  } else {
+    metadata->set_display_name(tag);
+  }
+  metadata->set_summary_description(image_description);
+
+  tensorflow::Summary::Image* image_summary = new tensorflow::Summary::Image();
+  image_summary->set_width(width);
+  image_summary->set_height(height);
+  image_summary->set_colorspace(channel);
+  image_summary->set_encoded_image_string(encoded_image);
+
+  tensorflow::Summary* summary = new tensorflow::Summary();
+  tensorflow::Summary::Value* summary_value = summary->add_value();
+  if (!summary_value) {
+    LOG(ERROR) << "[" << tag << "] Failed to add value to summary";
+    return;
+  }
+
+  summary_value->set_tag(tag);
+  summary_value->set_allocated_image(image_summary);
+  summary_value->set_allocated_metadata(metadata);
+
+  AddEvent(summary, step);
 }
 
+void TBoardFile::AddScalar(const std::string& tag, double value,
+                           uint32_t step) {
+  LOG(INFO) << "Adding scalar: " << tag << ": " << value << ", Step: " << value;
+
+  tensorflow::Summary* summary = new tensorflow::Summary();
```

Comment:
```suggestion
  tensorflow::Summary summary;
```
Protobufs are meant to be used as structures. In general, you will notice new/delete (or malloc/free) are not used in UchenML.

Reply:
The main point is that `Event` API expects a pointer to `summary`, and then `Event` will be responsible for deleting `summary` from memory. This may lead to multiple memory management failures eventually in my opinion, even Valgrind's detecting to many failure points. specially multiple destruction, internally inside `Event` and at the end of `TBoard` functions. 

But ok let's handle it this way and see what could happen, may be `Event` API is safe enough and have internal checkers.